Introduce mkBinaryCache function (#194345)
NixOS/nixpkgs: doc/builders/images/binarycache.section.md

# pkgs.mkBinaryCache {#sec-pkgs-binary-cache}

> `pkgs.mkBinaryCache` is a function for creating Nix flat-file binary caches. Such a cache exists as a directory on disk, and can be used as a Nix substituter by passing `--substituter file:///path/to/cache` to Nix commands.

Reviewer: Please make a full usage example, starting with such a derivation and using it on a full command line. While what you've written is in principle enough for an experienced user to abstract over, it is by far not detailed enough for a beginner to make sense of. Where does the binary cache end up? A full example will show the store path on stdout. How exactly do you use it as substituter? A full example will present a complete command line.

Also please link to the Nix manual for the relevant commands and flags if you mention them in the prose part, this makes discovery much easier.

Author: Okay, I've made the docs more descriptive. Could you take another look @fricklerhandwerk ?

> The parameters of `mkBinaryCache` with an example value are described below. This derivation will construct a flat-file binary cache containing the closure of `hello`.

Reviewer: ```suggestion
## Example
```

Use the description of how it works in the introduction.

```nix
mkBinaryCache {
  rootPaths = [hello];
}
```

> - `rootPaths` specifies a list of root derivations. The transitive closure of these derivations will be copied into the cache.

Reviewer: Use a definition list, they better match the semantics of what we want to achieve here.

```suggestion
`rootPaths` _(list of derivations)_

: List of root derivations. The transitive closure of these derivations will be copied into the cache.
```

Author: @fricklerhandwerk FWIW, no other documentation in the `images/` folder does this. They all use normal lists.

Reviewer: Yeah true, we don't have any standard yet and I think it would be worth writing it down. (I would if time was not severely limited and the list of "would be great to have" items was not growing without bounds.) 

Like all reviews here, this one about definition lists is just my personal opinion and I can't and don't want to tell anyone what to do. But I think more precise semantics is always better, and the cost here is negligible (in fact I think it's already a short term saving, because it already has more structure in the markdown source than unordered lists) and I try to keep it down for you offering a ready-made diff to pick up.

I'd rather have someone stumble over inconsistent formatting and possibly offer a fix than perpetuating a strictly inferior approach. Feel free to disagree. :)

Author: I would say "strictly inferior" is strong words when talking about a weird extended Markdown syntax that most people don't remember :P. And also implementing such changes in a scattershot way and hoping someone comes along later to fix the rest is the wrong way to do things; it should be in its own PR. I'd prefer not to be the person who messes up the consistency of the docs in this PR.

Reviewer: Markdown just happens to be the agreed-upon least terrible tool we have available. I fully understand your sentiment about introducing inconsistency. Do you have a proposal that has a high chance to actually be implemented? My intuition is, and I may well be wrong, that making changes incrementally while people are already at it, and discussing the merits, as we do now, even at the cost of temporary and potentially long-standing inconsistencies is much more probable to build up momentum for substantial change. Big Bang updates are risky and require much more dedication to pull through. Maybe I'm overestimating the total effort to convert all argument lists to a different format, for example, but there are many sweeping changes, not just in documentation, that many will welcome and still no one has done.

Anyway, just my two cents.

Author: I think I'll switch this PR back to a normal list if that's all right with you. I consider it bad form to introduce inconsistency. In the current world, switching to definition lists is an unrelated change and unrelated changes should not be piggybacked onto random PRs. I agree with you definition lists look slightly better when formatted, and I'd encourage you to write a PR to switch the docs over! Doing it in one fell swoop is infinitely preferable IMO because then there's a clear/unified place for discussion to happen, and a clear single commit to blame/revert if needed.

# pkgs.mkBinaryCache {#sec-pkgs-binary-cache}

`pkgs.mkBinaryCache` is a function for creating Nix flat-file binary caches. Such a cache exists as a directory on disk, and can be used as a Nix substituter by passing `--substituter file:///path/to/cache` to Nix commands.

> Nix packages are most commonly shared between machines using [HTTP, SSH, or S3](https://nixos.org/manual/nix/stable/package-management/sharing-packages.html), but a flat-file binary cache can still be useful in some situations. For example, you can copy it directly to another machine, or make it available on a network file system. It can also be a convenient way to make some Nix packages available inside a container via bind-mounting, when you want to efficiently run Nix builds inside a container without doing a writable mount of the host's `/nix` directory.

Reviewer 2: After reading this, I'd almost be convinced that this is the way to do binary caches on filesystems. It is not, and I would not recommend to use it except in very specific cases. It should at least say that the idiomatic way to work with an on-filesystem cache is through `nix-copy-closure` or `nix copy`, because it is more efficient and flexible that way.

> when you want to efficiently run Nix builds inside a container without doing a writable mount of the host's `/nix` directory.

`dockerTools` can already do this for you. It will put the packages in the local store rather than a separate directory, so no copying is necessary. This is arguably more efficient. What constraints might make the `dockerTools` solution infeasible for someone?

Author: > What constraints might make the dockerTools solution infeasible for someone?

That would require you to know in advance what you want baked into the store. The use-case of a bind-mounted binary cache is that you can mount different binary caches onto a fixed Docker image. Why you might want to do this is a bit of a rabbit hole, I think I'll just shorten this section to not mention it.

Author: Okay, I'm done with another pass on the docs. Please take a look!

> ## Example
> 
> The following derivation will construct a flat-file binary cache containing the closure of `hello`.
> 
> ```nix
> mkBinaryCache {
>   rootPaths = [hello];
> }
> ```
> 
> - `rootPaths` specifies a list of root derivations. The transitive closure of these derivations will be copied into the cache.

Reviewer: ````suggestion
## Arguments

- `rootPaths` specifies a list of root derivations. The transitive closure of these derivations will be copied into the cache.

## Examples

The following derivation will construct a flat-file binary cache containing the closure of `hello`.

```nix
mkBinaryCache {
  rootPaths = [hello];
}
```
````

Let's keep the examples together, I guess?

Author: Putting the arguments in their own section above is different from how all the other documentation does it; please let's keep it as-is and maintain consistency.

Reviewer: I agree in principle with consistency, but now you interleave examples and arguments. Without the sub-heading its not clear that there is an argument listing.

Reviewer: I totally feel you, the PR cycle can really be exhausting.

The suggestions are just suggestions, and we're trying to find the best solution. There is no authoritative stance on the topic, this is why there is so much case-by-case bickering.

My basis of argument is probably weak from your POV: just because it's done in a certain way it doesn't mean it's good. I'd rather start improving right now and do it in a principled manner on fresh material, define a rule while at it, and point people to it whenever they touch old material.

You don't have to agree, you own this code after all. Your approach is not worse in terms of results. The reason I don't favor it is that Big Bang changes are tedious to execute for a few reasons and a focused effort in that direction will never be prioritised because the overall payoff is small. That's the only reason I prefer temporary (even longer-term) inconsistency due to ongoing improvements over almost guaranteed permanent non-improvement.

Reviewer 2: > There is no authoritative stance on the topic

Is there work being done on guidelines?

Author: Well, putting aside the Big Bang debate on which you already know my position -- in this particular case I don't even agree that your alternative is better. I think it's more verbose, and doesn't flow as well since usually you want to know the details on a function argument after seeing how it's used.

I'd totally support an authoritative style guide though. Anyway, I couldn't quite tell from your message -- is that an insist or not insist? :)

Reviewer: > The suggestions are just suggestions

No insisting to be found here. :)

Author: Okay then, with apologies I'm going to leave it as-is and resolve this comment.

I might have said this before, but if you'd accept some advice -- I really don't think this is an appropriate or effective way to go about obtaining changes to the docs. If there's going to be a change to how the docs are written, then that deserves a full and public discussion process. At a minimum, the people who wrote the docs the way they currently are deserve a chance to weigh in. If I were a docs contributor and found that someone had forced through a convention change via drive-by on someone else's PR, without review, then I might be upset. Especially if it wasn't universally obvious that the change is an improvement.

On the positive side, you can be the one to drive the changes you want to see! The first obvious step would be to write the authoritative style guide and get it discussed and approved. And if that goes well, the next step would be to write incremental PRs to transform the docs. Maybe add a link to the style guide to the GitHub PR message, or make a git hook that fires when people touch the docs, or a even come up with a linter. That's the sort of process I'm accustomed to seeing on big open-source projects.

As always I appreciate the review and wish you luck on your documentation-improving missions.

Reviewer: Thanks for the feedback, I really appreciate it, and agree with everything you wrote in that regard. I do work on the guidelines, and I'm trying to take every stake holder on board with that. It's a constant struggle to strike the right balance between incremental progress and large-scale changes.

Here's an example of building and using the cache.

> ```bash
> # Build the cache on machine "host1"
> host1> nix-build -E 'with import <nixpkgs> {}; mkBinaryCache { rootPaths = [hello]; }'
> /nix/store/cc0562q828rnjqjyfj23d5q162gb424g-binary-cache
> 
> # Copy the resulting directory to "host2"
> host2> nix-build -A hello '<nixpkgs>' \
>   --option require-sigs false \
>   --option trusted-substituters file:///path/to/binary-cache \
>   --option substituters file:///path/to/binary-cache
> /nix/store/gl5a41azbpsadfkfmbilh9yk40dh5dl0-hello-2.12.1
> ```

Reviewer: I found the following form helpful for both reading the source and rendering the output appropriately. With the right CSS one can better distinguish inputs and outputs. In any case, prose should be prose and not code comments, unless there is no good way around it.

Also, please make sure the examples can actually be used in full. Don't use placeholders if it can be avoided.

````suggestion
Build the cache on one machine, `host1`:

```shellSession
nix-build -E 'with import <nixpkgs> {}; mkBinaryCache { rootPaths = [hello]; }'
```

    /nix/store/cc0562q828rnjqjyfj23d5q162gb424g-binary-cache

Copy the resulting directory to the other machine, `host2`:

```shellSession
scp result host2:/tmp/hello-cache
```

Build the derivation using the flat-file binary cache on the other machine, `host2`:
```shellSession
nix-build -A hello '<nixpkgs>' \
  --option require-sigs false \
  --option trusted-substituters file:///tmp/hello-cache \
  --option substituters file:///tmp/hello-cache
/nix/store/gl5a41azbpsadfkfmbilh9yk40dh5dl0-hello-2.12.1
```
```

Author: Okay, I applied this -- not sure why you're mixing in four-space code blocks since AFAIK those are identical to fenced code blocks. I'm going to change those back to normal fenced blocks.

# pkgs.mkBinaryCache {#sec-pkgs-binary-cache}

`pkgs.mkBinaryCache` is a function for creating Nix flat-file binary caches. Such a cache exists as a directory on disk, and can be used as a Nix substituter by passing `--substituter file:///path/to/cache` to Nix commands.

Nix packages are most commonly shared between machines using [HTTP, SSH, or S3](https://nixos.org/manual/nix/stable/package-management/sharing-packages.html), but a flat-file binary cache can still be useful in some situations. For example, you can copy it directly to another machine, or make it available on a network file system. It can also be a convenient way to make some Nix packages available inside a container via bind-mounting.

> Note that this function is meant for advanced use-cases. The more idiomatic way to work with flat-file binary caches is via the [nix copy](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-copy.html) command. You may also want to consider [dockerTools](#sec-pkgs-dockerTools) for your containerization needs.

Reviewer: There is an unwritten rule not to use experimental Nix features in Nixpkgs documentation, and while not everyone may agree, let's at least keep it consistent.

```suggestion
Note that this function is meant for advanced use-cases. The more idiomatic way to work with flat-file binary caches is via the [nix-copy-closure](https://nixos.org/manual/nix/stable/command-ref/nix-copy-closure.html) command. You may also want to consider [dockerTools](#sec-pkgs-dockerTools) for your containerization needs.
```

> Note that this function is meant for advanced use-cases. The more idiomatic way to work with flat-file binary caches is via the [nix-copy-closure](https://nixos.org/manual/nix/stable/command-ref/nix-copy-closure.html) command. You may also want to consider [dockerTools](#sec-pkgs-dockerTools) for your containerization needs.

Reviewer: ```suggestion
Note that this function is meant for advanced use-cases. The more idiomatic way to work with flat-file binary caches is via the [`nix-copy-closure`](https://nixos.org/manual/nix/stable/command-ref/nix-copy-closure.html) command. You may also want to consider [`dockerTools`](#sec-pkgs-dockerTools) for your containerization needs.
```

Author: This actually looks worse when rendered because the links no longer render as blue, so you can't easily tell it's a link.

Reviewer: That's a problem to be solved in CSS. The source should have rich semantics regardless of the representation du jour. Making things inline code tells the reader it's a code token and not some weird jargon term or other proper name.

> ## Example

Reviewer: ```suggestion
## Examples
```

Since we now have multiple in this section

Author: It's still a single example.

Reviewer: What about the example right below, which is not separated by a section header?

Reviewer: Right, it continues afterwards. I misread it as another example.

The following derivation will construct a flat-file binary cache containing the closure of `hello`.

```nix
mkBinaryCache {
  rootPaths = [hello];
}
```

> - `rootPaths` specifies a list of root derivations. The transitive closure of these derivations will be copied into the cache.

Reviewer 2: ```suggestion
- `rootPaths` specifies a list of root derivations. The transitive closure of these derivations' outputs will be copied into the cache.
```

Answers the question: why aren't the build dependencies copied to the cache?

> - `rootPaths` specifies a list of root derivations. The transitive closure of these derivations' outputs will be copied into the cache.

Reviewer: ```suggestion
`rootPaths` specifies a list of root derivations. The transitive closure of these derivations' outputs will be copied into the cache.
```

Since we don't have a proper argument listing, this may just as well be a regular example description. A list with one item seems pointless to me as long as we don't have a strict form for API docs.

Author: It's an argument listing like any other in these docs. As always, I want to maintain consistency.

Here's an example of building and using the cache.

Build the cache on one machine, `host1`:

```shellSession
nix-build -E 'with import <nixpkgs> {}; mkBinaryCache { rootPaths = [hello]; }'
```

> ```shellSession
> /nix/store/cc0562q828rnjqjyfj23d5q162gb424g-binary-cache
> ```

Reviewer: ````suggestion
```
/nix/store/cc0562q828rnjqjyfj23d5q162gb424g-binary-cache
```
````

That would be plain output, right? There is no consensus yet whether to format shell sessions as input output pairs as I proposed in the past, or continue relying on parsers for markup and prefixing the input lines with shell prompts. My preferred way, because it is the most readable in source and rendered representation, and also easiest to write:

````suggestion
    /nix/store/cc0562q828rnjqjyfj23d5q162gb424g-binary-cache
````

Author: Four spaces should be identical to a plain fenced code block when rendered (and is AFAICT). I favor the latter because it works for all text, even with newlines, and I don't like mixing two equivalent ways of doing things. Plain vs. `shellSession` makes no difference in rendering in this case.

Reviewer: Rendering doesn't matter, semantics do, and rendering is based on semantics.

Copy the resulting directory to the other machine, `host2`:

```shellSession
scp result host2:/tmp/hello-cache
```

> Build the derivation using the flat-file binary cache on the other machine, `host2`:

Author: Done

> Build the derivation using the flat-file binary cache on the other machine, `host2`:

Author: Done
